### PR TITLE
Fixes stack overflow in http inspector test

### PR DESCRIPTION
**Pull request by davinci26**
Signed-off-by: Sotiris Nanopoulos <sonanopo@microsoft.com>

Fixes stack overflow exception in `HttpInspectorTest.Http1WithLargeRequestLine` and makes the test faster. 

Additional Description:
While I was working on level vs edge based events I observed that the test is causing a stack overflow on Windows/MSVC.

The testlist http_inspector_test now passes on Windows but it should not because it relies on `Event::FileTriggerType::Edge` which are not supported. This is why I did not enable it for the CI.

Risk Level: N/A
Testing: N/A
Docs Changes: N/A
Release Notes: N/A

### CODE REVIEW

**Comment by davinci26** [comment]
I will look into the ASAN failure

**Comment by jmarantz** [comment]
ping when ready for review

/wait

File: test/extensions/filters/listener/http_inspector/BUILD [hunk 1]
```diff
@@ -15,6 +15,9 @@ envoy_extension_cc_test(
     name = "http_inspector_test",
     srcs = ["http_inspector_test.cc"],
     extension_name = "envoy.filters.listener.http_inspector",
+    #TODO(davinci26): The test passes on Windows *but* http inspector
+    # relies on Event::FileTriggerType::Edge and we get away with it
+    # because we mock the dispather.
```
Comment (jmarantz) [inline test/extensions/filters/listener/http_inspector/BUILD:20]: spelling: dispatcher.

File: test/extensions/filters/listener/http_inspector/http_inspector_test.cc [hunk 2]
```diff
@@ -584,19 +584,20 @@ TEST_F(HttpInspectorTest, Http1WithLargeRequestLine) {
...
-              }));
-    }
+    auto ctr = std::make_shared<size_t>(1);
+    EXPECT_CALL(os_sys_calls_, recv(42, _, _, MSG_PEEK))
+        .Times(num_loops)
+        .WillRepeatedly(Invoke([&data, ctr, num_loops](os_fd_t, void* buffer, size_t length,
+                                                       int) -> Api::SysCallSizeResult {
+          size_t len = (*ctr);
+          if (num_loops == 2) {
+            len = size_t(Config::MAX_INSPECT_SIZE / (3 - (*ctr)));
```
Comment (jmarantz) [inline test/extensions/filters/listener/http_inspector/http_inspector_test.cc:594]: `ASSERT(*ctx != 3)` here?

File: test/extensions/filters/listener/http_inspector/http_inspector_test.cc [hunk 3]
```diff
@@ -584,19 +584,20 @@ TEST_F(HttpInspectorTest, Http1WithLargeRequestLine) {
...
-                return Api::SysCallSizeResult{ssize_t(len), 0};
-              }));
-    }
+    auto ctr = std::make_shared<size_t>(1);
+    EXPECT_CALL(os_sys_calls_, recv(42, _, _, MSG_PEEK))
+        .Times(num_loops)
+        .WillRepeatedly(Invoke([&data, ctr, num_loops](os_fd_t, void* buffer, size_t length,
+                                                       int) -> Api::SysCallSizeResult {
+          size_t len = (*ctr);
+          if (num_loops == 2) {
```
Comment (jmarantz) [inline test/extensions/filters/listener/http_inspector/http_inspector_test.cc:593]: some comments here about what this is trying to accomplish would be good. It looks like maybe you are splitting up the buffer into a few chunks to mock streaming out some content?

File: test/extensions/filters/listener/http_inspector/http_inspector_test.cc [hunk 4]
```diff
@@ -584,19 +584,20 @@ TEST_F(HttpInspectorTest, Http1WithLargeRequestLine) {
...
+            len = size_t(Config::MAX_INSPECT_SIZE / (3 - (*ctr)));
+          }
+          ASSERT(length >= len);
+          memcpy(buffer, data.data(), len);
+          (*ctr) += 1;
```
Comment (jmarantz) [inline test/extensions/filters/listener/http_inspector/http_inspector_test.cc:598]: parens around `*ctr` not needed.

File: test/extensions/filters/listener/http_inspector/http_inspector_test.cc (hunk 3 above)
Comment (davinci26) [reply test/extensions/filters/listener/http_inspector/http_inspector_test.cc:593]: My understanding is that this test validates that in the case of a really large http request which is split over multiple reads the http listener is correctly parsing it.

File: test/extensions/filters/listener/http_inspector/BUILD (hunk 1 above)
Comment (davinci26) [reply test/extensions/filters/listener/http_inspector/BUILD]: Thanks good catch. I improved the comment overall in d540ffd472c7a3ceb9097f1b09b1d16acbf8b4bc. Now it takes into consideration that #12584 is merged

File: test/extensions/filters/listener/http_inspector/http_inspector_test.cc (hunk 2 above)
Comment (davinci26) [reply test/extensions/filters/listener/http_inspector/http_inspector_test.cc:598]: It should never happen but it is free to add the assert if someone ever tweaks the test. Added it in d540ffd472c7a3ceb9097f1b09b1d16acbf8b4bc

File: test/extensions/filters/listener/http_inspector/http_inspector_test.cc (hunk 4 above)
Comment (davinci26) [reply test/extensions/filters/listener/http_inspector/http_inspector_test.cc]: Fixed in d540ffd472c7a3ceb9097f1b09b1d16acbf8b4bc

File: test/extensions/filters/listener/http_inspector/http_inspector_test.cc (hunk 3 above)
Comment (davinci26) [reply test/extensions/filters/listener/http_inspector/http_inspector_test.cc:596]: Added a comment in d540ffd472c7a3ceb9097f1b09b1d16acbf8b4bc let me know if you want me to elaborate more

**Comment by davinci26** [comment]
The CI should go green now, sorry for the spam/notifications @jmarantz

**Comment by davinci26** [comment]
Picking up latest master to fix coverage issues.

**Pull request merged**